crates/rpc/rpc-types-compat/src/transaction/mod.rs

```diff
@@ -1,117 +1,104 @@
 //! Compatibility functions for rpc `Transaction` type.
 
-use alloy_primitives::{Address, TxKind};
+mod signature;
+mod typed;
+
+pub use signature::*;
+pub use typed::*;
+
+use std::fmt;
+
 use alloy_rpc_types::{
     request::{TransactionInput, TransactionRequest},
     TransactionInfo,
 };
-use reth_primitives::{TransactionSignedEcRecovered, TxType};
+use reth_primitives::{TransactionSigned, TransactionSignedEcRecovered, TxType};
 use reth_rpc_types::{Transaction, WithOtherFields};
-use signature::from_primitive_signature;
-pub use typed::*;
 
-mod signature;
-mod typed;
-
-/// Create a new rpc transaction result for a mined transaction, using the given [`TransactionInfo`]
-/// to populate the corresponding fields in the rpc result.
-pub fn from_recovered_with_block_context(
+/// Create a new rpc transaction result for a mined transaction, using the given block hash,
+/// number, and tx index fields to populate the corresponding fields in the rpc result.
+///
+/// The block hash, number, and tx index fields should be from the original block where the
+/// transaction was mined.
+pub fn from_recovered_with_block_context<T: TransactionCompat>(
     tx: TransactionSignedEcRecovered,
     tx_info: TransactionInfo,
-) -> WithOtherFields<Transaction> {
-    fill(tx, tx_info)
+) -> T::Transaction {
+    T::fill(tx, tx_info)
 }
 
 /// Create a new rpc transaction result for a _pending_ signed transaction, setting block
 /// environment related fields to `None`.
-pub fn from_recovered(tx: TransactionSignedEcRecovered) -> WithOtherFields<Transaction> {
-    fill(tx, TransactionInfo::default())
+pub fn from_recovered<T: TransactionCompat>(tx: TransactionSignedEcRecovered) -> T::Transaction {
+    T::fill(tx, TransactionInfo::default())
 }
 
-/// Create a new rpc transaction result for a _pending_ signed transaction, setting block
-/// environment related fields to `None`.
-fn fill(
-    tx: TransactionSignedEcRecovered,
-    tx_info: TransactionInfo,
-) -> WithOtherFields<Transaction> {
-    let signer = tx.signer();
-    let signed_tx = tx.into_signed();
+/// Builds RPC transaction w.r.t. network.
+pub trait TransactionCompat: Send + Sync + Unpin + Clone + fmt::Debug {
```

Comment:
I get why we currently need this.

but I worry that this could cause more trouble once we have network specific transaction types.

this is also not stateful, so these conversions are limited to the function input

Reply:
yes agreed, this is a first step, hence parallel work on https://github.com/alloy-rs/alloy/pull/1173, https://github.com/alloy-rs/alloy/issues/1315